Notifs post email functionality

ryanlieu opened the pull request:
### Summary of Changes

_Include a summary of changes then remove this line_

### Tests

_What tests did you add or modify and why? If no tests were added or modified, explain why. Remove this line_

### Documentation

_What documentation did you add or modify and why? Add any relevant links then remove this line_

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes, including screenshots of any UI changes. 
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public python functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes all tests documented in the [developer guide](https://github.com/lyft/amundsenfrontendlibrary/blob/master/docs/developer_guide.md#testing)
- [ ] I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"

ttannis commented on amundsen_application/api/utils/notification_utils.py:
```diff
@@ -0,0 +1,37 @@
+
+    TODO - Unify data format returned by search and metadata.
+    """
+    notification_type_dict = {
```
Let's protect ourselves from a bug that would happen if we try to access `notification_type_dict` with an invalid `notification_type`. We want to make sure we don't try to access the `'html'` key on an undefined variable. There's a couple ways to do this, you could write a helper method, or you could use `.get` and return a default value. i.e. `notification_type_dict.get(notification_type, {}).get('html')`

The second argument of `.get()` on python dictionaries is what will get returned if the value doesn't exist, else it  returns `None`.

ttannis commented on amundsen_application/static/templates/notification_added.html:
```diff
@@ -0,0 +1,3 @@
+<!doctype html>
```
nit: let's create a `templates/notifications` directory for these.

ttannis commented on amundsen_application/static/js/components/Feedback/index.tsx:
```diff
@@ -17,6 +17,8 @@ import {
 
 // TODO: Use css-modules instead of 'import'
 import './styles.scss';
+import { sendNotification } from 'ducks/notification/api/v0';
```
I see this is for testing, let's make sure to add a `TODO: Delete after xyz` where `xyz` is whatever task allows you test this out some other way, I assume it'll be when we have the first iteration of thee "Requesting" component. Same for the comment chunk added in `toggle`.

ttannis commented on amundsen_application/api/mail/v0.py:
```diff
@@ -83,7 +89,33 @@ def notification() -> Response:
+        notification_content = get_notification_content(data['notificationType'], data['options'])
+
+        response = mail_client.send_email(
+            recipients=data['recipients'],
```
We can add some safety here by using `get_query_param` from `request_utils`.  You can see examples of how we use it `metadata/v0`. We want to treat the React application as a black box (i.e. pretend we are not writing the logic or that we know the logic) to protect against the potential for something to go wrong there and the values don't get passed. 

If we use something like `get_query_param` (which could be renamed to something more applicable now but don't worry about that), if something goes wrong at least that method will raise an exception with a helpful error message about what parameter is missing.

ryanlieu commented:
@ttannis back up for review, should be good to look at again. Addressed comments, only large change outside of that was introducing a new exception type when the mail client is not implemented as well as removing typing for get_query_param (could return Dict, List, String, etc., not just String)

ttannis commented on amundsen_application/api/mail/v0.py:
```diff
@@ -7,24 +7,35 @@
-@mail_blueprint.route('/feedback', methods=['POST'])
-def feedback() -> Response:
-    """ An instance of BaseMailClient client must be configured on MAIL_CLIENT """
+class MailClientNotImplemented(Exception):
```
We have logic for raising a "not implemented" error in other places. How about we put this in a new `amundsen_application/api/exceptions.py` can call it `NotImplementedException`. After it's merged down to master we can then use it other places as well.

ryanlieu replied:
:+1:

ttannis commented on amundsen_application/api/mail/v0.py:
```diff
@@ -63,6 +74,10 @@ def feedback() -> Response:
             logging.error(message)
 
         return make_response(jsonify({'msg': message}), status_code)
+    except MailClientNotImplemented as e:
```
nit: I'm not sure if giving this the same alias `e` will cause any obscure python bugs in the future, but we could always follow the existing convention in our other repos of using `e1`, `e2`, etc

ryanlieu commented on amundsen_application/api/utils/request_utils.py:
```diff
@@ -6,7 +6,7 @@
 
 
-def get_query_param(args: Dict, param: str, error_msg: str = None) -> str:
+def get_query_param(args: Dict, param: str, error_msg: str = None):  # type: ignore
```
MyPy was giving a few errors as get_query_params was returning non-str objects that were valid

ttannis replied:
Ah just noticed your comment on this. I'm not a fan of loosening the typing on existing code. Perhaps the mistake was recommending that you use this method instead of creating a new one for your purposes given that your use case is different. 

Another option instead of using `get_query_param` could just be the simpler approach of using `.get()` and passing in some safe defaults. But how about I pose this instead -- what do you think will be best to do in the case where the React app has some issue and doesn't pass the right arguments? Let's agree on the behavior before I make any other recommendation.

ryanlieu replied:
I feel like if any data's malformed or it doesn't exist then an exception should be thrown and a 500 should be returned

ryanlieu replied:
@ttannis if that's the case, then directly accessing the dict would be fine as a KeyError would be thrown if the data is in fact malformed.

ttannis replied:
I see let's go with that then. Apologies for the bit of extra churn.

ryanlieu commented:
@ttannis should be good to look over again
